sdk/src/wallet/migration/migrate_0.rs

```diff
@@ -169,39 +136,205 @@ trait Convert {
+    );
+
+    string_serde_impl!(TransactionId);
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct Transaction {
+        pub payload: TransactionPayload,
+        pub block_id: Option<serde_json::Value>,
+        pub inclusion_state: InclusionState,
+        pub timestamp: u128,
+        pub transaction_id: TransactionId,
+        pub network_id: u64,
+        pub incoming: bool,
+        pub note: Option<String>,
+        #[serde(default)]
+        pub inputs: Vec<OutputWithMetadataResponse>,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    pub struct TransactionPayload {
+        pub essence: TransactionEssence,
+        pub unlocks: serde_json::Value,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(tag = "type", content = "data")]
+    pub enum TransactionEssence {
+        Regular(RegularTransactionEssence),
+    }
+
+    #[derive(Serialize, Deserialize)]
+    pub struct RegularTransactionEssence {
+        pub network_id: u64,
+        pub inputs: serde_json::Value,
+        pub inputs_commitment: serde_json::Value,
+        pub outputs: serde_json::Value,
+        pub payload: serde_json::Value,
+    }
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct OutputWithMetadataResponse {
+        pub metadata: OutputMetadataDto,
+        pub output: serde_json::Value,
+    }
+
+    pub struct OutputId {
+        pub transaction_id: TransactionId,
+        pub index: u16,
+    }
+
+    impl OutputId {
+        pub const LENGTH: usize = TransactionId::LENGTH + core::mem::size_of::<u16>();
+    }
+
+    impl TryFrom<[u8; Self::LENGTH]> for OutputId {
+        type Error = Error;
+
+        fn try_from(bytes: [u8; Self::LENGTH]) -> Result<Self, Self::Error> {
+            let (transaction_id, index) = bytes.split_at(TransactionId::LENGTH);
+
+            Ok(Self {
+                transaction_id: From::<[u8; TransactionId::LENGTH]>::from(transaction_id.try_into().unwrap()),
+                index: u16::from_le_bytes(index.try_into().unwrap()),
+            })
+        }
+    }
+
+    impl FromStr for OutputId {
+        type Err = Error;
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            Self::try_from(prefix_hex::decode::<[u8; Self::LENGTH]>(s).map_err(Error::Hex)?)
+        }
+    }
+
+    impl core::fmt::Display for OutputId {
+        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+            let mut buffer = [0u8; Self::LENGTH];
+            let (transaction_id, index) = buffer.split_at_mut(TransactionId::LENGTH);
+            transaction_id.copy_from_slice(self.transaction_id.as_ref());
+            index.copy_from_slice(&self.index.to_le_bytes());
+            write!(f, "{}", prefix_hex::encode(buffer))
+        }
+    }
+
+    string_serde_impl!(OutputId);
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct OutputMetadata {
```

Comment:
Should we maybe add a version suffix or have own types for each version?
Since this is a new type, it could be used in a future migration as the old type and then it would be wasted to duplicate it
Without any versioning it's probably hard to keep track of what was being used where

Reply:
but aren't those types essentially versioned by living inside `migrate_0` module? The `types` module should probably be `pub(crate)` though to have access to them from other `migrate_x` modules.

Reply:
> but aren't those types essentially versioned by living inside migrate_0 module?
If you just look at OutputMetadata or OutputMetadataDto you don't know which one is the old or new version

Reply:
I don't think it's a good idea to share the internal types between migrations.

Reply:
Why not if they're the same?

Reply:
Sure, if they are the same then it's fine. We can always make it more public in the future.

Reply:
> > but aren't those types essentially versioned by living inside migrate_0 module?
> 
> If you just look at OutputMetadata or OutputMetadataDto you don't know which one is the old or new version

that's true ... I personally would just fully qualify in that case, but I can see your point. Since it's just migration types I don't care about the type name, so I'ld be fine with the suffix tbh.

Reply:
Yeah alright, fine if we change when really needed